src/lib/wallet/GoodWallet.js

```diff
@@ -312,7 +311,7 @@ export class GoodWallet {
       receipt: tx
     }
   }
-
+  //FIXME: what's this for? why does it read events from block0
```

Comment:
This is from an old PR, but it was thought to look for the sender's address and also to check the status of the retreat (for example: if it is already done, if it exists).

Instead of searching from the `0` block, we could search from the block in which the contract was deployed to save time; but at the time of doing so, I found no other way to search for the sender's address.

Reply:
I still don't understand. You have all info to check canWithdraw from public views in the smart contract.
You can check if the hash holds value by checking the payments mapping. hash -> amount. if its 0 then it has been withdrawn or never assigned value.
Should I add a helper method?

Reply:
`isLinkUsed` is a public view. I think that by checking it, we can give the user a better reason why the link isn't valid.

`payments` is public and was the way I found to check for the amount available to be withdrawn for that particular hash.

Looking for the events was my way to find who generated the link as the `sender` property didn't exist at that time.

Reply:
changed to use senders...